src/lib/support/SafeInt.h

```diff
@@ -105,6 +105,38 @@ bool CanCastTo(U arg)
     return 0 <= arg && static_cast<uintmax_t>(arg) <= static_cast<uintmax_t>(numeric_limits<T>::max());
 }
 
+/**
+ * A function to reverse the effects of a signed-to-unsigned integer cast.
+ *
+ * If the argument is small enough to be representable as a positive signed
+ * integer, returns that integer.  Otherwise, returns a negative integer which
+ * would, if cast to the type of the argument, produce the given value.
+ *
+ * So for example, if a uint8_t with value 254 is passed in this function will
+ * return an int8_t with value -2.
+ */
+template <typename T>
+typename std::enable_if<std::is_unsigned<T>::value, typename std::make_signed<T>::type>::type CastToSigned(T arg)
```

Comment:
Could we have some unit tests for this? They would serve as examples as well.

Reply:
Yes, will add tests.